Sources/Authenticator/Authenticator.swift

```diff
@@ -0,0 +1,81 @@
+//
+//  Authenticator.swift
+//  Comet
+//
+//  Created by Tuan Tu Do on 26.05.2021.
+//  Copyright © 2021 Etnetera. All rights reserved.
+//
+
+import Combine
+import Foundation
+
+final class Authenticator {
+    private let tokenProvider: TokenProviding
+    private let queue = DispatchQueue(label: "Authenticator-\(UUID().uuidString)")
+    private var refreshTokenPublisher: AnyPublisher<String, AuthenticatorError>?
+
+    init(tokenProvider: TokenProviding) {
+        self.tokenProvider = tokenProvider
+    }
+
+    var accessToken: AnyPublisher<String, AuthenticatorError> {
+        queue.sync { [weak self] in
+            guard let unwrappedSelf = self else {
+                return Fail(error: AuthenticatorError.internalError).eraseToAnyPublisher()
+            }
+
+            if let publisher = unwrappedSelf.refreshTokenPublisher {
+                return publisher
+            }
+
+            return unwrappedSelf.tokenProvider.accessToken
+                .mapError { $0.authenticatorError }
+                .eraseToAnyPublisher()
+        }
+    }
+
+    var refreshAccessToken: AnyPublisher<String, AuthenticatorError> {
+        queue.sync { [weak self] in
+            guard let unwrappedSelf = self else {
+                return Fail(error: AuthenticatorError.internalError).eraseToAnyPublisher()
+            }
+
+            if let publisher = unwrappedSelf.refreshTokenPublisher {
+                return publisher
+            }
+
+            let publisher = unwrappedSelf.tokenProvider.refreshAccessToken
+                .mapError { $0.authenticatorError }
+                .handleEvents(receiveCompletion: { _ in
+                    unwrappedSelf.refreshTokenPublisher = nil
+                })
+                .share()
+                .eraseToAnyPublisher()
```

Comment:
As discussed: There is a synchronization issue here.

Reply:
Fixed with `shareReplay` operator.